ganttproject/src/main/java/net/sourceforge/ganttproject/storage/SqlProjectDatabaseImpl.kt

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2022 BarD Software s.r.o., Anastasiia Postnikova
+
+This file is part of GanttProject, an open-source project management tool.
+
+GanttProject is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+GanttProject is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with GanttProject.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+package net.sourceforge.ganttproject.storage
+
+import net.sourceforge.ganttproject.storage.Tables.*
+import net.sourceforge.ganttproject.GPLogger
+import net.sourceforge.ganttproject.task.Task
+import org.h2.jdbcx.JdbcDataSource
+import org.jooq.DSLContext
+import org.jooq.SQLDialect
+import org.jooq.impl.DSL
+import org.jooq.impl.DSL.primaryKey
+import org.jooq.impl.SQLDataType.*
+import java.sql.SQLException
+import javax.sql.DataSource
+
+class SqlProjectDatabaseException(message: String) : ProjectDatabaseException(message)
+
+class SqlProjectDatabaseImpl internal constructor(private val dataSource: DataSource): ProjectDatabase {
+  companion object Factory {
+    fun createInMemoryDatabase(): ProjectDatabase {
+      val dataSource = JdbcDataSource()
+      dataSource.setURL(H2_IN_MEMORY_URL)
+      return SqlProjectDatabaseImpl(dataSource)
+    }
+  }
+
+  private fun <T> withExecutor(
+    errorMessage: String = "Failed to execute query",
+    body: (executor: DSLContext) -> T
```

Comment:
also, minor nitpicking: the name "executor" in JVM apps is usually taken by ExecutorService and its friends, and GP is no exception. jOOQ named this object "DSLContext", so perhaps we can call it "dsl" too